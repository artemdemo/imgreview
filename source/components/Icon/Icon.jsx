import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import waitForFontAwesome from './waitForFontAwesome';

import './Icon.less';

/**
 * Icon component
 * @param props
 * @link https://fontawesome.com/get-started
 */
class Icon extends React.PureComponent {
    constructor(props) {
        super(props);

        this.state = {
            fontLoaded: false,
        };
    }

    componentDidMount() {
        waitForFontAwesome
            .then(() => {
                this.setState({
                    fontLoaded: true,
                });
            })
            .catch(() => {
                console.warn('Error while loading font awesome');
            });
    }

    render() {
        const { name, inText, className, title } = this.props;

        if (name === '' || name == null) {
            throw new Error('Icon prop `name` couldn\'t be empty');
        }

        const iconClass = classnames(className, 'fa', {
            [`fa-${name}`]: true,
            'icon_in-text': inText,
        });

        if (this.state.fontLoaded) {
            return (
                <span
                    className={iconClass}
                    title={title}
                />
            );
        }

        return (
            <span>{title}</span>
        );
    }
}

Icon.propTypes = {
    name: PropTypes.string.isRequired,
    className: PropTypes.string,
    title: PropTypes.string,
    inText: PropTypes.bool,
};

Icon.defaultProps = {
    className: '',
    title: null,
    inText: false,
};

export default Icon;
