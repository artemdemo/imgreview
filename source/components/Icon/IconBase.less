@import '~font-awesome/css/font-awesome.min.css';

.icon_in-text {
    padding-right: 5px;
}
