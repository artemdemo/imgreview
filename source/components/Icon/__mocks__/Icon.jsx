import React from 'react';

const Icon = props => (
    <div data-mock='Icon'>
        {props.children}
    </div>
);

export default Icon;
