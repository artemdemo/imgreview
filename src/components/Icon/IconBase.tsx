import React from 'react';
import classnames from 'classnames';
import styled from 'styled-components';
import waitForFontAwesome from './waitForFontAwesome';

import 'font-awesome/css/font-awesome.min.css';

export type TIconProps = {
    name: string;
    className?: string;
    title?: string;
    inText?: boolean;
};

type State = {
    fontLoaded: boolean;
};

const IconSty = styled.span<{inText?: boolean}>`
    padding-right: ${props => props.inText ? '5px' : '0'};
`;

/**
 * Icon component
 * @param props
 * @link https://fontawesome.com/get-started
 */
class Icon extends React.PureComponent<TIconProps, State> {
    private fontAwesomePromise: Promise<any>;
    private unmounted: boolean = false;

    static readonly defaultProps = {
        className: '',
        title: null,
        inText: false,
    };

    constructor(props) {
        super(props);

        this.state = {
            fontLoaded: false,
        };
    }

    componentDidMount() {
        this.fontAwesomePromise = waitForFontAwesome
            .then(() => {
                if (!this.unmounted) {
                    this.setState({
                        fontLoaded: true,
                    });
                }
            })
            .catch(() => {
                console.warn('Error while loading font awesome');
            });
    }

    componentWillUnmount() {
        this.unmounted = true;
    }

    render() {
        const { name, inText, className, title } = this.props;

        if (name === '' || name == null) {
            throw new Error('Icon prop `name` couldn\'t be empty');
        }

        const iconClass = classnames(className, 'fa', {
            [`fa-${name}`]: true,
        });

        if (this.state.fontLoaded) {
            return (
                <IconSty
                    className={iconClass}
                    title={title}
                    inText={inText}
                />
            );
        }

        return (
            <IconSty
                inText={inText}
            >
                {title}
            </IconSty>
        );
    }
}

export default Icon;
