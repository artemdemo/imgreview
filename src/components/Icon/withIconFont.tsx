import React from 'react';
import waitForFontAwesome from './waitForFontAwesome';

type TProps = {
    title?: string;
    name?: string;
    className?: string;
};

type TState = {
    fontLoaded: boolean;
};

const withIconFont = Component => {
    class ComponentWrapper extends React.PureComponent<TProps, TState> {
        #unmounted: boolean = false;

        static readonly defaultProps = {
            title: undefined,
        };

        state = {
            fontLoaded: false,
        };

        componentDidMount() {
            waitForFontAwesome
                .then(() => {
                    if (!this.#unmounted) {
                        this.setState({
                            fontLoaded: true,
                        });
                    }
                })
                .catch(() => {
                    console.warn('Error while loading font awesome');
                });
        }

        componentWillUnmount() {
            this.#unmounted = true;
        }

        render() {
            if (this.state.fontLoaded) {
                return (
                    <Component {...this.props} />
                )
            }

            return this.props.title || '';
        }
    }

    return ComponentWrapper;
};

export default withIconFont;
