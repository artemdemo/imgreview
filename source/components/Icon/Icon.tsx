import React from 'react';
import { TIconProps } from './IconBase';

/**
 * Notice: This file has `.tsx` extension.
 * I need it only to support JSX code in mock file.
 * Typescript doesn't allow to write jsx code in `.ts` files, only in `.tsx`
 */
import lazify from '../../containers/lazify';

const LazyIcon = lazify(() => import(/* webpackChunkName: "<IconBase />" */ './IconBase'));

const IconPlaceholder = (props: TIconProps) => {
    return (
        <LazyIcon
            {...props}
        >
            {props.title}
        </LazyIcon>
    );
};

export default IconPlaceholder;
