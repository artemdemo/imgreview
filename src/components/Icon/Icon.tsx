import React from 'react';
import classnames from 'classnames';
import styled from 'styled-components';
import withIconFont from './withIconFont';

import 'font-awesome/css/font-awesome.min.css';

export type TIconProps = {
    name: string;
    className?: string;
    title?: string;
    inText?: boolean;
};

type TState = {
    fontLoaded: boolean;
};

const IconSty = styled.span<{inText?: boolean}>`
    padding-right: ${props => props.inText ? '5px' : '0'};
`;

/**
 * Icon component
 * @param props
 * @link https://fontawesome.com/get-started
 */
class Icon extends React.PureComponent<TIconProps, TState> {
    static readonly defaultProps = {
        className: '',
        title: null,
        inText: false,
    };

    render() {
        const { name, inText, className, title } = this.props;

        if (name === '' || name == null) {
            throw new Error('Icon prop `name` couldn\'t be empty');
        }

        const iconClass = classnames(className, 'fa', {
            [`fa-${name}`]: true,
        });

        return (
            <IconSty
                className={iconClass}
                title={title}
                inText={inText}
            />
        );
    }
}

export default withIconFont(Icon);
