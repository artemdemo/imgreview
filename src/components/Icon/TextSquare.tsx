import React from 'react';
import styled from 'styled-components';
import Icon from './Icon';
import withIconFont from './withIconFont';

const Stack = styled.span`
    zoom: 0.7;
    top: -5px;
    width: 1.6em;
    height: 1.1em;
    line-height: 1.7em;
    vertical-align: unset;
`;

type TProps = {
    title: string;
};

const TextSquare = (props: TProps) => {
    return (
        <Stack className='fa-stack' title={props.title}>
            <Icon name='square-o' className='fa-stack-2x' />
            <Icon name='font' className='fa-stack-1x' />
        </Stack>
    );
};

export default withIconFont(TextSquare);
