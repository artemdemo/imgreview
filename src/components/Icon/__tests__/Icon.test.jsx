/* eslint-disable new-cap */
import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Icon from '../Icon';

jest.mock('../waitForFontAwesome.ts');

describe('Icon', () => {
    it('should render with name', () => {
        const wrapper = mount(
            <Icon name='fire' />,
        );
        wrapper.setState({
            fontLoaded: true,
        });

        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should render with name and className', () => {
        const wrapper = mount(
            <Icon name='fire' className='some-class' />,
        );
        wrapper.setState({
            fontLoaded: true,
        });

        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should render with name and title', () => {
        const wrapper = mount(
            <Icon name='fire' title='Fire' />,
        );
        wrapper.setState({
            fontLoaded: true,
        });

        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should render with name and inText', () => {
        const wrapper = mount(
            <Icon name='fire' inText />,
        );
        wrapper.setState({
            fontLoaded: true,
        });

        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should throw an error if no name provided', () => {
        expect(() => {
            const icon = new Icon();
            icon.props = {
                name: '',
            };
            icon.render();
        }).toThrow();

        expect(() => {
            const icon = new Icon();
            icon.props = {
                name: null,
            };
            icon.render();
        }).toThrow();

        expect(() => {
            const icon = new Icon();
            icon.props = {
                name: undefined,
            };
            icon.render();
        }).toThrow();
    });
});
