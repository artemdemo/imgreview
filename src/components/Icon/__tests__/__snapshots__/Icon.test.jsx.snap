// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon should render with name 1`] = `
<Icon
  className=""
  inText={false}
  name="fire"
  title={null}
>
  <styled.span
    className="fa fa-fire"
    inText={false}
    title={null}
  >
    <span
      className="sc-AxjAm hEFEr fa fa-fire"
      title={null}
    />
  </styled.span>
</Icon>
`;

exports[`Icon should render with name and className 1`] = `
<Icon
  className="some-class"
  inText={false}
  name="fire"
  title={null}
>
  <styled.span
    className="some-class fa fa-fire"
    inText={false}
    title={null}
  >
    <span
      className="sc-AxjAm hEFEr some-class fa fa-fire"
      title={null}
    />
  </styled.span>
</Icon>
`;

exports[`Icon should render with name and inText 1`] = `
<Icon
  className=""
  inText={true}
  name="fire"
  title={null}
>
  <styled.span
    className="fa fa-fire"
    inText={true}
    title={null}
  >
    <span
      className="sc-AxjAm gMDvjq fa fa-fire"
      title={null}
    />
  </styled.span>
</Icon>
`;

exports[`Icon should render with name and title 1`] = `
<Icon
  className=""
  inText={false}
  name="fire"
  title="Fire"
>
  <styled.span
    className="fa fa-fire"
    inText={false}
    title="Fire"
  >
    <span
      className="sc-AxjAm hEFEr fa fa-fire"
      title="Fire"
    />
  </styled.span>
</Icon>
`;
